src/source/PeerConnection/SessionDescription.h

```diff
@@ -45,11 +46,15 @@ extern "C" {
 #define DEFAULT_PAYLOAD_OPUS    (UINT64) 111
 #define DEFAULT_PAYLOAD_VP8     (UINT64) 96
 #define DEFAULT_PAYLOAD_H264    (UINT64) 125
+#define DEFAULT_PAYLOAD_H265    (UINT64) 127
 
 #define DEFAULT_PAYLOAD_MULAW_STR (PCHAR) "0"
 #define DEFAULT_PAYLOAD_ALAW_STR  (PCHAR) "8"
 
-#define DEFAULT_H264_FMTP   (PCHAR) "level-asymmetry-allowed=1;packetization-mode=1;profile-level-id=42e01f"
+#define DEFAULT_H264_FMTP (PCHAR) "level-asymmetry-allowed=1;packetization-mode=1;profile-level-id=42e01f"
+#define DEFAULT_H265_FMTP                                                                                                                            \
```

Comment:
Need to discuss if we want to restrict the profile and other properties. Mostly inclined towards removing this but wanted everyone's opinion here

Reply:
webrtc is only mandated to support I think the constrained baseline profile, removing this would allow frames to be sent which are not part of webrtc spec.  I suspect we will need. specific FMTP line with properties/profile for H265 as well

Reply:
There is a draft RFC for H265 support in WebRTC:  https://datatracker.ietf.org/doc/draft-ietf-avtcore-hevc-webrtc/

It is only a draft but probably best to follow it for now, relevant section is:

```
   For the H.265 [H.265] codec, endpoints MUST support the payload
   formats defined in [RFC7798]. In addition, they MUST support Main
   Profile Level 3.1 (level-id=93) and SHOULD support Main Profile Level
   4 (level-id=120).
```

We need to make sure we have tested with these and we have space in our design to separate between different profiles as needed.

Reply:
I tested this with the Android SDK and I don't receive an fmtp in the offer from libwebrtc:
```
2024-04-25 05:48:35.312 DEBUG   setRemoteDescription(): REMOTE_SDP:v=0
o=- 842804224343444729 2 IN IP4 127.0.0.1
s=-
t=0 0
a=group:BUNDLE 0 1 2
a=msid-semantic: WMS
m=audio 9 UDP/TLS/RTP/SAVPF 96 0 8 105 13 113 126
c=IN IP4 0.0.0.0
a=rtcp:9 IN IP4 0.0.0.0
a=ice-ufrag:wcqN
a=ice-pwd:v+PR8YjabtyDlJyDEMPSg3xW
a=ice-options:trickle renomination
a=fingerprint:sha-256 F3:4E:6E:A9:07:F4:C4:16:56:90:EE:5F:0C:C9:DD:80:64:3D:3B:59:24:7C:22:29:8C:FB:7D:BF:80:C4:26:4F
a=setup:actpass
a=mid:0
a=extmap:1 urn:ietf:params:rtp-hdrext:ssrc-audio-level
a=extmap:2 http://www.webrtc.org/experiments/rtp-hdrext/abs-send-time
a=extmap:3 http://www.ietf.org/id/draft-holmer-rmcat-transport-wide-cc-extensions-01
a=extmap:4 urn:ietf:params:rtp-hdrext:sdes:mid
a=extmap:5 urn:ietf:params:rtp-hdrext:sdes:rtp-stream-id
a=extmap:6 urn:ietf:params:rtp-hdrext:sdes:repaired-rtp-stream-id
a=recvonly
a=rtcp-mux
a=rtpmap:96 AAC/16000
a=rtpmap:0 PCMU/8000
a=rtpmap:8 PCMA/8000
a=rtpmap:105 CN/16000
a=rtpmap:13 CN/8000
a=rtpmap:113 telephone-event/16000
a=rtpmap:126 telephone-event/8000
m=video 9 UDP/TLS/RTP/SAVPF 127 97 98 99 100 101 102 103 104 125 106
c=IN IP4 0.0.0.0
a=rtcp:9 IN IP4 0.0.0.0
a=ice-ufrag:wcqN
a=ice-pwd:v+PR8YjabtyDlJyDEMPSg3xW
a=ice-options:trickle renomination
a=fingerprint:sha-256 F3:4E:6E:A9:07:F4:C4:16:56:90:EE:5F:0C:C9:DD:80:64:3D:3B:59:24:7C:22:29:8C:FB:7D:BF:80:C4:26:4F
a=setup:actpass
a=mid:1
a=extmap:14 urn:ietf:params:rtp-hdrext:toffset
a=extmap:2 http://www.webrtc.org/experiments/rtp-hdrext/abs-send-time
a=extmap:13 urn:3gpp:video-orientation
a=extmap:3 http://www.ietf.org/id/draft-holmer-rmcat-transport-wide-cc-extensions-01
a=extmap:12 http://www.webrtc.org/experiments/rtp-hdrext/playout-delay
a=extmap:11 http://www.webrtc.org/experiments/rtp-hdrext/video-content-type
a=extmap:7 http://www.webrtc.org/experiments/rtp-hdrext/video-timing
a=extmap:8 http://tools.ietf.org/html/draft-ietf-avtext-framemarking-07
a=extmap:9 http://www.webrtc.org/experiments/rtp-hdrext/color-space
a=extmap:4 urn:ietf:params:rtp-hdrext:sdes:mid
a=extmap:5 urn:ietf:params:rtp-hdrext:sdes:rtp-stream-id
a=extmap:6 urn:ietf:params:rtp-hdrext:sdes:repaired-rtp-stream-id
a=recvonly
a=rtcp-mux
a=rtcp-rsize
a=rtpmap:127 VP8/90000
a=rtcp-fb:127 goog-remb
a=rtcp-fb:127 transport-cc
a=rtcp-fb:127 ccm fir
a=rtcp-fb:127 nack
a=rtcp-fb:127 nack pli
a=rtpmap:97 rtx/90000
a=fmtp:97 apt=127
a=rtpmap:98 VP9/90000
a=rtcp-fb:98 goog-remb
a=rtcp-fb:98 transport-cc
a=rtcp-fb:98 ccm fir
a=rtcp-fb:98 nack
a=rtcp-fb:98 nack pli
a=rtpmap:99 rtx/90000
a=fmtp:99 apt=98
a=rtpmap:100 H264/90000
a=rtcp-fb:100 goog-remb
a=rtcp-fb:100 transport-cc
a=rtcp-fb:100 ccm fir
a=rtcp-fb:100 nack
a=rtcp-fb:100 nack pli
a=fmtp:100 level-asymmetry-allowed=1;packetization-mode=1;profile-level-id=42e01f
a=rtpmap:101 rtx/90000
a=fmtp:101 apt=100
a=rtpmap:102 H265/90000
a=rtcp-fb:102 goog-remb
a=rtcp-fb:102 transport-cc
a=rtcp-fb:102 ccm fir
a=rtcp-fb:102 nack
a=rtcp-fb:102 nack pli
a=rtpmap:103 rtx/90000
a=fmtp:103 apt=102
a=rtpmap:104 red/90000
a=rtpmap:125 rtx/90000
a=fmtp:125 apt=104
a=rtpmap:106 ulpfec/90000
m=application 9 UDP/DTLS/SCTP webrtc-datachannel
c=IN IP4 0.0.0.0
a=ice-ufrag:wcqN
a=ice-pwd:v+PR8YjabtyDlJyDEMPSg3xW
a=ice-options:trickle renomination
a=fingerprint:sha-256 F3:4E:6E:A9:07:F4:C4:16:56:90:EE:5F:0C:C9:DD:80:64:3D:3B:59:24:7C:22:29:8C:FB:7D:BF:80:C4:26:4F
a=setup:actpass
a=mid:2
a=sctp-port:5000
a=max-message-size:262144

```
Should we be following the draft still?

Reply:
Safari's offer with H265 doesn't have an fmtp at 104 (h265/90000) either:
```
[Log] PeerConnectionBackend::createOfferSucceeded(CCD61DEB87F60F75) Create offer succeeded:
v=0
o=- 985372420936776446 2 IN IP4 127.0.0.1
s=-
t=0 0
a=group:BUNDLE 0 1 2
a=extmap-allow-mixed
a=msid-semantic: WMS bc57d7ff-a695-4aa3-b860-7de2e4494fed
m=audio 9 UDP/TLS/RTP/SAVPF 111 63 9 0 8 13 110 126
c=IN IP4 0.0.0.0
a=rtcp:9 IN IP4 0.0.0.0
a=ice-ufrag://pY
a=ice-pwd:svCwcEunmp3xe5js1YGHMWI8
a=ice-options:trickle
a=fingerprint:sha-256 64:61:0D:9D:DE:D0:37:5A:82:55:73:E9:41:06:30:79:C9:31:B0:5B:89:80:99:9D:B6:E9:61:F2:5D:1E:83:42
a=setup:actpass
a=mid:0
a=extmap:1 urn:ietf:params:rtp-hdrext:ssrc-audio-level
a=extmap:2 http://www.webrtc.org/experiments/rtp-hdrext/abs-send-time
a=extmap:3 http://www.ietf.org/id/draft-holmer-rmcat-transport-wide-cc-extensions-01
a=extmap:4 urn:ietf:params:rtp-hdrext:sdes:mid
a=sendrecv
a=msid:bc57d7ff-a695-4aa3-b860-7de2e4494fed 833dba34-201d-4710-b1ce-8273bf20a245
a=rtcp-mux
a=rtpmap:111 opus/48000/2
a=rtcp-fb:111 transport-cc
a=fmtp:111 minptime=10;useinbandfec=1
a=rtpmap:63 red/48000/2
a=fmtp:63 111/111
a=rtpmap:9 G722/8000
a=rtpmap:0 PCMU/8000
a=rtpmap:8 PCMA/8000
a=rtpmap:13 CN/8000
a=rtpmap:110 telephone-event/48000
a=rtpmap:126 telephone-event/8000
a=ssrc:2776556564 cname:VmWUiGQTNzJgTWxC
a=ssrc:2776556564 msid:bc57d7ff-a695-4aa3-b860-7de2e4494fed 833dba34-201d-4710-b1ce-8273bf20a245
m=video 9 UDP/TLS/RTP/SAVPF 96 97 98 99 100 101 102 103 104 105 106 107 108 109 127 125 112 113 114
c=IN IP4 0.0.0.0
a=rtcp:9 IN IP4 0.0.0.0
a=ice-ufrag://pY
a=ice-pwd:svCwcEunmp3xe5js1YGHMWI8
a=ice-options:trickle
a=fingerprint:sha-256 64:61:0D:9D:DE:D0:37:5A:82:55:73:E9:41:06:30:79:C9:31:B0:5B:89:80:99:9D:B6:E9:61:F2:5D:1E:83:42
a=setup:actpass
a=mid:1
a=extmap:14 urn:ietf:params:rtp-hdrext:toffset
a=extmap:2 http://www.webrtc.org/experiments/rtp-hdrext/abs-send-time
a=extmap:13 urn:3gpp:video-orientation
a=extmap:3 http://www.ietf.org/id/draft-holmer-rmcat-transport-wide-cc-extensions-01
a=extmap:5 http://www.webrtc.org/experiments/rtp-hdrext/playout-delay
a=extmap:6 http://www.webrtc.org/experiments/rtp-hdrext/video-content-type
a=extmap:7 http://www.webrtc.org/experiments/rtp-hdrext/video-timing
a=extmap:8 http://www.webrtc.org/experiments/rtp-hdrext/color-space
a=extmap:4 urn:ietf:params:rtp-hdrext:sdes:mid
a=extmap:10 urn:ietf:params:rtp-hdrext:sdes:rtp-stream-id
a=extmap:11 urn:ietf:params:rtp-hdrext:sdes:repaired-rtp-stream-id
a=sendrecv
a=msid:bc57d7ff-a695-4aa3-b860-7de2e4494fed 094be066-b91c-4de0-adbe-b47bcf534826
a=rtcp-mux
a=rtcp-rsize
a=rtpmap:96 H264/90000
a=rtcp-fb:96 goog-remb
a=rtcp-fb:96 transport-cc
a=rtcp-fb:96 ccm fir
a=rtcp-fb:96 nack
a=rtcp-fb:96 nack pli
a=fmtp:96 level-asymmetry-allowed=1;packetization-mode=1;profile-level-id=640c1f
a=rtpmap:97 rtx/90000
a=fmtp:97 apt=96
a=rtpmap:98 H264/90000
a=rtcp-fb:98 goog-remb
a=rtcp-fb:98 transport-cc
a=rtcp-fb:98 ccm fir
a=rtcp-fb:98 nack
a=rtcp-fb:98 nack pli
a=fmtp:98 level-asymmetry-allowed=1;packetization-mode=1;profile-level-id=42e01f
a=rtpmap:99 rtx/90000
a=fmtp:99 apt=98
a=rtpmap:100 H264/90000
a=rtcp-fb:100 goog-remb
a=rtcp-fb:100 transport-cc
a=rtcp-fb:100 ccm fir
a=rtcp-fb:100 nack
a=rtcp-fb:100 nack pli
a=fmtp:100 level-asymmetry-allowed=1;packetization-mode=0;profile-level-id=640c1f
a=rtpmap:101 rtx/90000
a=fmtp:101 apt=100
a=rtpmap:102 H264/90000
a=rtcp-fb:102 goog-remb
a=rtcp-fb:102 transport-cc
a=rtcp-fb:102 ccm fir
a=rtcp-fb:102 nack
a=rtcp-fb:102 nack pli
a=fmtp:102 level-asymmetry-allowed=1;packetization-mode=0;profile-level-id=42e01f
a=rtpmap:103 rtx/90000
a=fmtp:103 apt=102
a=rtpmap:104 H265/90000
a=rtcp-fb:104 goog-remb
a=rtcp-fb:104 transport-cc
a=rtcp-fb:104 ccm fir
a=rtcp-fb:104 nack
a=rtcp-fb:104 nack pli
a=rtpmap:105 rtx/90000
a=fmtp:105 apt=104
a=rtpmap:106 VP8/90000
a=rtcp-fb:106 goog-remb
a=rtcp-fb:106 transport-cc
a=rtcp-fb:106 ccm fir
a=rtcp-fb:106 nack
a=rtcp-fb:106 nack pli
a=rtpmap:107 rtx/90000
a=fmtp:107 apt=106
a=rtpmap:108 VP9/90000
a=rtcp-fb:108 goog-remb
a=rtcp-fb:108 transport-cc
a=rtcp-fb:108 ccm fir
a=rtcp-fb:108 nack
a=rtcp-fb:108 nack pli
a=fmtp:108 profile-id=0
a=rtpmap:109 rtx/90000
a=fmtp:109 apt=108
a=rtpmap:127 VP9/90000
a=rtcp-fb:127 goog-remb
a=rtcp-fb:127 transport-cc
a=rtcp-fb:127 ccm fir
a=rtcp-fb:127 nack
a=rtcp-fb:127 nack pli
a=fmtp:127 profile-id=2
a=rtpmap:125 rtx/90000
a=fmtp:125 apt=127
a=rtpmap:112 red/90000
a=rtpmap:113 rtx/90000
a=fmtp:113 apt=112
a=rtpmap:114 ulpfec/90000
a=ssrc-group:FID 3773084415 3113708224
a=ssrc:3773084415 cname:VmWUiGQTNzJgTWxC
a=ssrc:3773084415 msid:bc57d7ff-a695-4aa3-b860-7de2e4494fed 094be066-b91c-4de0-adbe-b47bcf534826
a=ssrc:3113708224 cname:VmWUiGQTNzJgTWxC
a=ssrc:3113708224 msid:bc57d7ff-a695-4aa3-b860-7de2e4494fed 094be066-b91c-4de0-adbe-b47bcf534826
m=application 9 UDP/DTLS/SCTP webrtc-datachannel
c=IN IP4 0.0.0.0
a=ice-ufrag://pY
a=ice-pwd:svCwcEunmp3xe5js1YGHMWI8
a=ice-options:trickle
a=fingerprint:sha-256 64:61:0D:9D:DE:D0:37:5A:82:55:73:E9:41:06:30:79:C9:31:B0:5B:89:80:99:9D:B6:E9:61:F2:5D:1E:83:42
a=setup:actpass
a=mid:2
a=sctp-port:5000
a=max-message-size:262144
```

Reply:
I verified the behavior for using fmtp. We send the default fmtp only if we are sending the offer ourselves first. If not, we emulate the behavior in the offer. If the offer doesn't contain fmtp for h265, we don't include it in the answer either. If it does, we include the same fmtp that we received in the answer. I think it should be fine to keep this in that case. However, we do not assume the fmtp specified here is our default profile and that is why the name is RTC_CODEC_H265 as opposed to H264's codec name

Reply:
Thanks for verifying the behavior.